lua/entities/gmod_wire_expression2/core/player.lua

```diff
@@ -676,3 +676,179 @@ end
+	for ex,_ in pairs(DeathAlert) do -- loops over all chips in deathalert, ignores key.
+		if IsValid(ex) then
+			ex.context.data.runByDeath = entry
+			ex:Execute()
+			ex.context.data.runByDeath = nil
+		end
+	end
+end)
+
+hook.Add("PlayerSpawn","Exp2PlayerDetRespn",function(ply,transition)
+	local entry = DEFAULT_TABLE() -- default table
+	entry.s["Player"]=ply
+	entry.s["Timestamp"]=CurTime()
+	entry.stypes["Player"]="e"
+	entry.stypes["Timestamp"]="n"
+	entry.n[1]=ply
+	entry.n[2]=CurTime()
+	entry.ntypes[1]="e"
+	entry.ntypes[2]="n"
+	entry.size=4
+	RespawnList[ply:EntIndex()] = entry
+	RespawnList.last = entry
+	for ex,_ in pairs(RespawnAlert) do
+		if IsValid(ex) then
+			ex.context.data.runByRespawned = entry
+			ex:Execute()
+			ex.context.data.runByRespawned = nil
+		end
+	end
+end)
+
+__e2setcost(5)
+
+--- If active is 0, the chip will no longer run on death.
+e2function void runOnDeath(number activate)
+	if activate ~= 0 then
+		DeathAlert[self.entity] = true
+	else
+		DeathAlert[self.entity] = nil
+	end
+end
+
+e2function void runOnSpawn(number activate)
+	if activate ~= 0 then
+		RespawnAlert[self.entity] = true
+	else
+		RespawnAlert[self.entity] = nil
+	end
+end
+
+--If ran by death, (defined in e2 data), gives 1 or 0 (ternary)
+e2function number deathClk()
+	return self.data.runByDeath and 1 or 0
+end
+
+e2function number spawnClk()
+	return self.data.runByRespawned and 1 or 0
+end
+
+e2function number lastDeathTime() -- returns when the last death happened
+	local lastD = DeathList.last
+	return CHECK_VALIDT(lastD) and lastD.s["Timestamp"] or 0 -- Checks if table is valid, if so then returns the timestamp from the table, else returns 0.
+end
+
+e2function number lastDeathTime(entity ply) -- returns when the player provided last died
+	if not CHECK_VALIDP(ply) then return NULL end
+	local lastD = DeathList[ply:EntIndex()]
+	return CHECK_VALIDT(lastD) and lastD.s["Timestamp"] or 0 -- Checks if table is valid, if so then returns the timestamp from the table, else returns 0.
+end
+
+e2function number lastSpawnTime()
+	local lastS = RespawnList.last
+	return CHECK_VALIDT(lastS) and lastS.s["Timestamp"] or 0
+end
+
+e2function number lastSpawnTime(entity ply) -- returns the last time player provided spawned.
+	if not CHECK_VALIDP(ply) then return NULL end
+	local lastS = RespawnList.last
+	return CHECK_VALIDT(lastS) and lastS.s["Timestamp"] or 0
+end
+
+e2function table lastDeath(entity ply)
+	if not CHECK_VALIDP(ply) then return NULL end
+	local lastD = RespawnList[ply:EntIndex()]
+	return CHECK_VALIDT(lastD) and lastD or DEFAULT_TABLE()
+end
+
+e2function table lastSpawn(entity ply)
+	if not CHECK_VALIDP(ply) then return NULL end
+	local lastS = RespawnList[ply:EntIndex()]
+	return CHECK_VALIDT(lastS) and lastS or DEFAULT_TABLE()
+end
+
+e2function table lastDeath()
+	local lastD = DeathList.last
+	return CHECK_VALIDT(lastD) and lastD or DEFAULT_TABLE()
```

Comment:
I just realized earlier that since tables are pass by reference in lua, you can get this table and then modify it and then no one else on the server will know who killed who. Same thing for the function below this one. You'll need to copy the table, either by using `table.Copy` or by manually creating a new table and moving the values over. The latter would be slightly faster since you always know which values exist.

Reply:
Yeah.. another reason not to return table/array.